geomstats/geometry/sasaki_metric.py

```diff
@@ -0,0 +1,236 @@
+    x = x_ini
+    for _ in range(max_iter):
+        grad_x = grad(x)
+        grad_norm = gs.linalg.norm(grad_x)
+        if grad_norm < tol:
+            break
+        grad_x = -lrate * grad_x
+        x = exp(grad_x, x)
+    return x
+
+
+class SasakiMetric(RiemannianMetric):
+    """Implements of the Sasaki metric on the tangent bundle TM of a Riem. manifold M.
+
+    The Sasaki metric is characterized by the following three properties:
+     * the canonical projection of TM becomes a Riemannian submersion,
+     * parallel vector fields along curves are orthogonal to their fibres, and
+     * its restriction to any tangent space is Euclidean.
+
+    Geodesic computations are realized via a discrete formulation of the geodesic
+    equation on TM that involve geodesics, parallel translation, and the curvature
+    tensor on the base manifold M (see [1] for details).
+
+    Parameters
+    ----------
+    metric : RiemannianMetric
+        Metric of the base manifold of the tangent bundle.
+
+    References
+    ----------
+    .. [1] Muralidharan, P., & Fletcher, P. T. (2012, June).
+    Sasaki metrics for analysis of longitudinal data on manifolds.
+    In 2012 IEEE conference on computer vision and pattern recognition (pp. 1027-1034).
+    https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4270017/
+    """
+
+    def __init__(self, metric: RiemannianMetric):
+        self.metric = metric
+        shape = (2, gs.prod(gs.array(metric.shape)))
+
+        super(SasakiMetric, self).__init__(2 * metric.dim, shape=shape,
+                                           default_point_type='matrix')
+
+    def exp(self, tangent_vec, base_point, n_steps=N_STEPS, **kwargs):
+        """Compute the Riemannian exponential of a point.
+
+        Exponential map at base_point of tangent_vec computed by
+        shooting a Sasaki geodesic using an Euler integration on TM.
+
+        Parameters
+        ----------
+        tangent_vec : array-like, shape=[..., 2, M.dim]
+            Tangent vector in TTM at the base point in TM.
+        base_point : array-like, shape=[..., 2, M.dim]
+            Point in the tangent bundle TM of manifold M.
+        n_steps : int
+            Number of discrete time steps.
+            Optional, default: N_STEPS.
+
+        Returns
+        -------
+        exp : array-like, shape=[..., 2, M.dim]
+            Point on the tangent bundle TM.
+        """
+        bs_pts = gs.reshape(base_point, (-1, 2) + self.metric.shape)
+        tngs = gs.reshape(tangent_vec, bs_pts.shape)
+
+        metric = self.metric
+        par_trans = metric.parallel_transport
+        eps = 1 / n_steps
+
+        v0, w0 = gs.take(tngs, 0, axis=1), gs.take(tngs, 1, axis=1)
+        p0, u0 = gs.take(bs_pts, 0, axis=1), gs.take(bs_pts, 1, axis=1)
+        for _ in range(n_steps):
+            p = metric.exp(eps * v0, p0)
+            u = par_trans(u0 + eps * w0, p0, None, p)
```

Comment:
for clarity, can you change `par_trans(u0 + eps * w0, p0, None, p)` to `par_trans(u0 + eps * w0, p0, end_point=p)` in all the places we use it? I find it more readable.

Reply:
Done

Reply:
Done